lib/useOnyx.ts

```diff
@@ -0,0 +1,171 @@
+import {deepEqual} from 'fast-equals';
+import {useCallback, useEffect, useRef, useSyncExternalStore} from 'react';
+import Onyx from './Onyx';
+import type {CollectionKeyBase, KeyValueMapping, OnyxCollection, OnyxEntry, OnyxKey, Selector} from './types';
+import useLiveRef from './useLiveRef';
+import usePrevious from './usePrevious';
+
+type OnyxValue<TKey extends OnyxKey> = TKey extends CollectionKeyBase ? OnyxCollection<KeyValueMapping[TKey]> : OnyxEntry<KeyValueMapping[TKey]>;
+
+type UseOnyxOptions<TKey extends OnyxKey, TReturnData> = {
+    /**
+     * Determines if this key in this subscription is safe to be evicted.
+     */
+    canEvict?: boolean;
+
+    /**
+     * If set to false, then no data will be prefilled into the component.
+     */
+    initWithStoredValues?: boolean;
+
+    /**
+     * If set to true, data will be retrieved from cache during the first render even if there is a pending merge for the key.
+     */
+    allowStaleData?: boolean;
+
+    /**
+     * This value will be returned by the hook on the first render while the data is being read from Onyx.
+     */
+    initialValue?: OnyxValue<TKey>;
+
+    /**
+     * This will be used to subscribe to a subset of an Onyx key's data.
+     * Using this setting on `useOnyx` can have very positive performance benefits because the component will only re-render
+     * when the subset of data changes. Otherwise, any change of data on any property would normally
+     * cause the component to re-render (and that can be expensive from a performance standpoint).
+     */
+    selector?: Selector<TKey, unknown, TReturnData>;
+};
+
+type FetchStatus = 'loading' | 'loaded';
+
+type CachedValue<TKey extends OnyxKey, TValue> = TValue extends OnyxValue<TKey> ? TValue : TKey extends CollectionKeyBase ? NonNullable<OnyxCollection<TValue>> : TValue;
+
+type UseOnyxData<TKey extends OnyxKey, TValue> = {
```

Comment:
Hmmm, I thought we had aligned on this as a better pattern for `useOnyx` return data that's the best of both worlds:

- simple to get the "main" return value
- easy to extend with more configs / extra fields

```suggestion
type UseOnyxData<TKey extends OnyxKey, TValue> = [CachedValue<TKey, TValue>, {status: FetchStatus}];
```

that way we can do something like this:

```ts
const [account] = useOnyx(ONYXKEYS.ACCOUNT)l
const [reports, {status: reportsLoadingStatus}] = useOnyx(ONYXKEYS.COLLECTION.REPORT);
```